Windowsv3

anumang opened the pull request:
Windows UWP Support implemented.

dantman commented on index.js:
```diff
@@ -51,19 +51,37 @@ function pick(opts) {
-    audio: Platform.OS === 'ios' ? 'public.audio' : 'audio/*',
-    pdf: Platform.OS === 'ios' ? 'com.adobe.pdf' : 'application/pdf',
-  };
+  static types = PlatformTypes[Platform.OS]
```
This will make `types` undefined on unknown platforms. And I don't like tying the types list to platforms as the list grows.

It would be better if the Types object was `mimeTypes`, `uti`, and `extensions`. There was a separate platform -> {mimeTypes,uti,extensions} and the default if the platform is not known is `mimeTypes`.

dantman commented on windows/RNDocumentPicker/RCTDocumentPickerModule.cs:
```diff
@@ -0,0 +1,188 @@
+            }
+            else
+            {
+                promise.Reject(E_DOCUMENT_PICKER_CANCELED, E_DOCUMENT_PICKER_CANCELED);
```
Use the same `"User canceled document picker"` message as I did in Android instead of duplicating the key.

dantman commented on windows/RNDocumentPicker/RCTDocumentPickerModule.cs:
```diff
@@ -0,0 +1,188 @@
+                    }
+                    catch (Exception)
+                    {
+                        promise.Reject(E_FAILED_TO_SHOW_PICKER, E_FAILED_TO_SHOW_PICKER);
```
Get the error message from the actual exception instead of hiding it from the developer.

dantman commented on windows/RNDocumentPicker/RCTDocumentPickerModule.cs:
```diff
@@ -0,0 +1,188 @@
+                });
+            }
+            catch (Exception) {
+                promise.Reject(E_UNEXPECTED_EXCEPTION, E_UNEXPECTED_EXCEPTION);
```
Same here.

dantman commented on windows/RNDocumentPicker/RCTDocumentPickerModule.cs:
```diff
@@ -0,0 +1,188 @@
+            }
+            else
+            {
+                promise.Reject(E_DOCUMENT_PICKER_CANCELED, E_DOCUMENT_PICKER_CANCELED);
```
Same.

anumang commented:
Requested changes are fixed:
* Error messages improved
* Mime type selection is added like as Android for Windows

dantman commented:
I didn't want the Windows code to switch to trying to guess extensions based on mime types. The library especially should not be using a hardcoded type map.

anumang commented:
Maybe I misunderstood your point. If you explain it, I can try to handle.
Here is my understanding and if you also evaluate this, I will be appreciate:
"This will make types undefined on unknown platforms. And I don't like tying the types list to platforms as the list grows." 
--> I understood that we should use generic mime types instead of types list. At this point, Except IOS, Android & Windows & Other OS's we can use generic MimeTypes. This will be much flexible at future, like user can control pickers with desired mime types. As a result, not bounded to pre-defined type utils.

"It would be better if the Types object was mimeTypes, uti, and extensions. There was a separate platform -> {mimeTypes,uti,extensions} and the default if the platform is not known is mimeTypes."
--> I inspected several node & nuget libraries (mime-db, node-mime etc.) all of them using some of standard libraries as a hard coded type maps from;
* apache - Apache common media types
* iana - IANA-defined media types
* nginx - nginx media types
At this point, for MimeType utils, we bounded to hardcoded type maps. So I used same approach and used only nginx library because of seems much familiar with Windows OS.

dantman commented:
The Windows/C# code should still use extensions, that is what Windows expects for filters.

IMHO the JS code should have something like:
```js
const Types = {
  mimeTypes: {
     ...
  },
  utis: {...},
  extensions: {...},
};
```

Instead of a fixed PlatformTypes map, because if we ever get support for a 4th platform we're almost guaranteed that they will use one of these 3 formats we already have (likely mime types or extensions) instead of a 4th proprietary type filter like Apple does.

Then map platforms to those:
```js
const PlatformTypes = {
  android: Types.mimeTypes,
  ios: Types.utis,
  windows: Types.extensions,
};
```

And when you do the PlatformTypes map make sure that we fallback to mimeTypes on unknown platforms instead of `undefined`, `PlatformTypes[Platform.OS] || Types.mimeTypes`.

anumang commented:
I understood now and it seems better. If you find another improvement, I can fix it too.

dantman commented:
Looks good. Though the final thing @Elyx0 or I probably need to do is actually setting up a windows dummy project and trying it out.

This would be a great 3.1.0 feature release on its own.

Elyx0 commented:
Damn I'm one month late on this. I sooo don't have time 😱

anumang commented:
TY, but I want to say that I added one more option flag capability after you reviewed because of necessity:
 I saw that on UWP applications, user can pick a file anywhere but when processing this file later on app with URL, causing permission issues because of windows. To solve it, I added one more flag so this enables to cache copy file from picker.
This feature can be used only cache=true option so default behavior is same.
If cache flag is set, when user selecting a file or files, I am copying this files to app cache folder and return info of cached copy by,
```
var fileInCache = await file.CopyAsync(ApplicationData.Current.TemporaryFolder, file.Name.ToString(), NameCollisionOption.ReplaceExisting).AsTask().ConfigureAwait(false);
```
